tools/bridge/bridge/confirmation_sender.py

```diff
@@ -0,0 +1,105 @@
+from eth_keys.datatypes import PrivateKey
+from eth_utils import keccak, int_to_big_endian
+
+from bridge.constants import HOME_CHAIN_STEP_INTERVAL
+
+
+class ConfirmationSender:
+    """Sends confirmTransfer transactions to the home bridge contract."""
+
+    def __init__(
+        self,
+        transfer_queue: Queue,
+        home_bridge_contract: Contract,
+        private_key: bytes,
+        gas_price: int,
+        max_reorg_depth: int,
+    ):
+        self.transfer_queue = transfer_queue
+        self.home_bridge_contract = home_bridge_contract
+        self.private_key = private_key
+        self.address = PrivateKey(self.private_key).public_key.to_canonical_address()
+        self.gas_price = gas_price
+        self.max_reorg_depth = max_reorg_depth
+
+        self.w3 = self.home_bridge_contract.web3
+
+        self.logger = logging.getLogger("bridge.confirmation_sender.ConfirmationSender")
+
+        self.pending_transaction_queue: Queue[Dict[str, Any]] = Queue()
+
+    def get_next_nonce(self):
+        return self.w3.eth.getTransactionCount(self.address)
+
+    def run(self):
+        self.logger.info("Starting")
+        try:
+            greenlets = [
+                gevent.spawn(self.watch_pending_transactions),
+                gevent.spawn(self.send_confirmation_transactions),
+            ]
+            gevent.joinall(greenlets)
+        finally:
+            for greenlet in greenlets:
+                greenlet.kill()
+
+    def send_confirmation_transactions(self):
+        while True:
+            transfer = self.transfer_queue.get()
+            transaction = self.prepare_confirmation_transaction(transfer)
+            self.send_confirmation_transaction(transaction)
+
+    def prepare_confirmation_transaction(self, transfer):
+        nonce = self.get_next_nonce()
+        transaction = self.home_bridge_contract.functions.confirmTransfer(
+            self.compute_transfer_hash(transfer),
+            transfer.transactionHash,
+            transfer.args.value,
+            transfer.args["from"],
+        ).buildTransaction({"gasPrice": self.gas_price, "nonce": nonce})
+        signed_transaction = self.w3.eth.account.sign_transaction(
+            transaction, self.private_key
+        )
+        return signed_transaction
+
+    def compute_transfer_hash(self, transfer):
+        return keccak(
+            b"".join(
+                [bytes(transfer.transactionHash), int_to_big_endian(transfer.logIndex)]
+            )
+        )
+
+    def send_confirmation_transaction(self, transaction):
+        self.logger.info(f"Sending confirmation transaction {transaction}")
+        self.pending_transaction_queue.put(transaction)
+        self.w3.eth.sendRawTransaction(transaction.rawTransaction)
+
+    def watch_pending_transactions(self):
+        while True:
+            self.clear_confirmed_transactions()
+            gevent.sleep(HOME_CHAIN_STEP_INTERVAL)
+
+    def clear_confirmed_transactions(self):
```

Comment:
Question: I'm a bit confused. First this isn't a requirement for trustlines-network/project#583 as I see it. You didn't referenced a specific story task, but I thought this is the one you implemented. I guess this is not bad in first place, but it just blew up this whole PR. But possibly it is necessary.
Second I fail to see the purpose of this function. Apparently nobody else does, so I'm confused. What does this else then log a message? I mean a successful confirmation will be logged anyways by the related contract event. From this point of perspective this is a big overhead. What I could image is, that it is a first step into the direction of handle stuck transactions. Thanks for helping me!

Reply:
Main motivation for the pending transaction queue is to be able to resend transactions if they aren't confirmed (because too little gas). If a transaction is confirmed we need to remove them from the queue, that's what this method is for.

It's true that it isn't part of the requirements. I originally implemented it to be able to choose the nonce based on the transactions that are pending already (but then removed it because it's a short cut anyway and it got harder to do when I switched from a list to a queue.

Reply:
Okay I missed the early part of this PR. That's a good point. Thank you for the good explanation. :+1: